junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/AutoCloseTests.java

```diff
@@ -0,0 +1,191 @@
+		String msg = "@AutoClose: Cannot resolve the destroy method close() at AutoCloseNoCloseMethodFailingTestCase.resource: String";
+
+		Events tests = executeTestsForClass(AutoCloseNoCloseMethodFailingTestCase.class).testEvents();
+		assertFailingWithMessage(tests, msg);
+	}
+
+	@Test
+	void noShutdownMethod() {
+		String msg = "@AutoClose: Cannot resolve the destroy method shutdown() at AutoCloseNoShutdownMethodFailingTestCase.resource: String";
+
+		Events tests = executeTestsForClass(AutoCloseNoShutdownMethodFailingTestCase.class).testEvents();
+		assertFailingWithMessage(tests, msg);
+	}
+
+	@Test
+	void namespace() {
+		assertEquals(Namespace.create(AutoClose.class), AutoCloseExtension.NAMESPACE);
+	}
+
+	@Test
+	void spyPermitsOnlyASingleAction() {
+		AutoCloseSpy spy = new AutoCloseSpy("");
+
+		spy.close();
+
+		assertThrows(IllegalStateException.class, spy::close);
+		assertThrows(IllegalStateException.class, spy::run);
+		assertEquals(asList("close()"), recorder);
+	}
+
+	private static void assertFailingWithMessage(Events testEvent, String msg) {
+		testEvent.assertStatistics(stats -> stats.failed(1)).assertThatEvents().haveExactly(1,
+			finishedWithFailure(cause(message(actual -> actual.contains(msg)))));
+	}
+
+	static class AutoCloseTestCase {
+
+		private static @AutoClose AutoCloseable staticClosable;
+		private static @AutoClose AutoCloseable nullStatic;
+
+		private final @AutoClose AutoCloseable closable = new AutoCloseSpy("afterEach-");
+		private final @AutoClose("run") Runnable runnable = new AutoCloseSpy("afterEach-");
+		private @AutoClose AutoCloseable nullField;
+
+		@Test
+		void justPass() {
+			assertFields();
+		}
+
+		@Test
+		void anotherPass() {
+			assertFields();
+		}
+
+		private void assertFields() {
+			assertNotNull(staticClosable);
+			assertNull(nullStatic);
+
+			assertNotNull(closable);
+			assertNotNull(runnable);
+			assertNull(nullField);
+		}
+
+		@BeforeAll
+		static void setup() {
+			staticClosable = new AutoCloseSpy("afterAll-");
+		}
+
+	}
+
+	static class AutoCloseNoCloseMethodFailingTestCase {
+
+		@AutoClose
+		private final String resource = "nothing to close()";
+
+		@Test
+		void alwaysPass() {
+			assertNotNull(resource);
+		}
+
+	}
+
+	static class AutoCloseNoShutdownMethodFailingTestCase {
+
+		@AutoClose("shutdown")
+		private final String resource = "nothing to shutdown()";
+
+		@Test
+		void alwaysPass() {
+			assertNotNull(resource);
+		}
+
+	}
+
+	static class AutoCloseSpy implements AutoCloseable, Runnable {
+
+		private final String prefix;
+		private String invokedMethod = "";
+
+		public AutoCloseSpy(String prefix) {
+			this.prefix = prefix;
+		}
+
+		@Override
+		public void run() {
+			checkIfAlreadyInvoked();
+			recordInvocation("run()");
+		}
+
+		@Override
+		public void close() {
+			checkIfAlreadyInvoked();
+			recordInvocation("close()");
+		}
+
+		private void checkIfAlreadyInvoked() {
+			if (!invokedMethod.isEmpty())
+				throw new IllegalStateException();
```

Comment:
We do not use if-statements without blocks (`{...}`) in JUnit 5.

Reply:
Applied. But...
What about single return? 😁 
```
if (invokedMethod.isEmpty()) return;
throw new IllegalStateException();
```

Reply:
Nope, we don't do that either.

Reply:
TBH, for stuff like that we often use `Preconditions` to achieve the goal with a one-liner.